Sources/LanguageServerProtocol/Connection.swift

```diff
@@ -136,3 +137,30 @@ extension LocalConnection: Connection {
     return id
   }
 }
+
+extension Connection {
+  /// Send the given request to the connection and await its result.
+  ///
+  /// This method automatically sends a `CancelRequestNotification` to the
+  /// connection if the task it is executing in is being cancelled.
+  ///
+  /// - Warning: Because this message is `async`, it does not provide any ordering
+  ///   guarantees. If you need to gurantee that messages are sent in-order
+  ///   use the version with a completion handler.
+  public func send<R: RequestType>(_ request: R) async throws -> R.Response {
+    let requestIDWrapper = ThreadSafeBox<RequestID?>(initialValue: nil)
+    try Task.checkCancellation()
+    return try await withTaskCancellationHandler {
```

Comment:
Should the `checkCancellation` happen inside `withTaskCancellationHandler` to match the sourcekitd logic? Seems like there could be a potential race condition where the task is cancelled between the calls to `checkCancellation` and `withTaskCancellationHandler`, which would still lead to the request being sent

Reply:
Come to think of it, is there a potential race condition if the task is cancelled before `requestIDWrapper` is assigned? 🤔

Reply:
I think it doesn’t really matter in practice but I moved it for consistency.

> Come to think of it, is there a potential race condition if the task is cancelled before `requestIDWrapper` is assigned? 🤔

Yes, I think there is a potential race that could happen here. But since the window in which it can happen is really short and this is only cancellation, which is best-effort anyway, I decided that it’s OK to accept it. The alternative would be yet another dispatch queue that blocks `onCancel` from executing until the request ID has been set and that seemed a little bit like overkill to me.

Reply:
How would we actually cancel before the client has the request ID back anyway?

Reply:
In LSP, the client specifies the request ID

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#contentPart

Reply:
Ah I see, unfortunate.

Reply:
> The alternative would be yet another dispatch queue that blocks `onCancel` from executing until the request ID has been set and that seemed a little bit like overkill to me

I guess one other alternative would be to check again for cancellation afterwards, e.g something like:

```swift
extension ThreadSafeBox {
  func takeValue<U>() -> T where U? == T {
    lock.withLock {
      guard let value = self.value else { return nil }
      self.value = nil
      return value
    }
  }
}

func withCancellableThrowingContinuation<Success, CancelToken>(
  _ body: (CheckedContinuation<Success, any Error>) -> CancelToken,
  onCancel cancel: (CancelToken) -> Void
) async throws -> Success {
  let cancelWrapper = ThreadSafeBox<CancelToken?>(initialValue: nil)

  @Sendable
  func cancelFn() {
    if let token = cancelWrapper.takeValue() {
      cancel(token)
    }
  }

  return try await withTaskCancellationHandler(operation: {
    try Task.checkCancellation()
    return try await withCheckedThrowingContinuation { cont in
      cancelWrapper.value = body(cont)
      if Task.isCancelled {
        cancelFn()
      }
    }
  }, onCancel: cancelFn)
}
```

Not sure if that's worth it though.